From ef59fea2552db957a05ef01f4105b7ab9d1ab0a2 Mon Sep 17 00:00:00 2001
From: Martin Kletzander <mkletzan@redhat.com>
Date: Tue, 16 Jul 2013 14:32:34 +0200
Subject: [RHEL-6.5 virt-manager PATCH 7/8] Add support for security relabeling

https://bugzilla.redhat.com/show_bug.cgi?id=907399

virt-install already supports relabeling, but virt-manager doesn't and
in some cases, this can cause problems, for example when switching to
dynamic labeling with the relabeling turned off.  I took the approach
of allowing the user to choose, with safe fallbacks to defaults.

Deals also with this:

https://bugzilla.redhat.com/show_bug.cgi?id=907390
(cherry picked from commit b6f0ba364f6376793989450869db83e117586cec)

Conflicts:
	src/vmm-details.glade -- upstream commit 103c2e0da1

Signed-off-by: Martin Kletzander <mkletzan@redhat.com>
---
 src/virtManager/details.py | 24 ++++++++++++++++++------
 src/virtManager/domain.py  | 24 +++++++++++++++++++-----
 src/vmm-details.glade      | 18 +++++++++++++++++-
 3 files changed, 54 insertions(+), 12 deletions(-)

diff --git a/src/virtManager/details.py b/src/virtManager/details.py
index ee01d60..39dcc28 100644
--- a/src/virtManager/details.py
+++ b/src/virtManager/details.py
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2006-2008 Red Hat, Inc.
+# Copyright (C) 2006-2008, 2013 Red Hat, Inc.
 # Copyright (C) 2006 Daniel P. Berrange <berrange@redhat.com>
 #
 # This program is free software; you can redistribute it and/or modify
@@ -366,6 +366,7 @@ class vmmDetails(vmmGObjectUI):
             "on_overview_apic_changed": self.config_apic_changed,
             "on_overview_clock_changed": (self.enable_apply, EDIT_CLOCK),
             "on_security_label_changed": (self.enable_apply, EDIT_SECURITY),
+            "on_security_relabel_changed": (self.enable_apply, EDIT_SECURITY),
             "on_security_type_changed": self.security_type_changed,

             "on_config_vcpus_changed": self.config_vcpus_changed,
@@ -717,7 +718,7 @@ class vmmDetails(vmmGObjectUI):

         # Security info tooltips
         util.tooltip_wrapper(self.widget("security-static-info"),
-            _("Static SELinux security type tells libvirt to always start the guest process with the specified label. The administrator is responsible for making sure the images are labeled correctly on disk."))
+            _("Static SELinux security type tells libvirt to always start the guest process with the specified label. Unless 'relabel' is set, the administrator is responsible for making sure the images are labeled correctly on disk."))
         util.tooltip_wrapper(self.widget("security-dynamic-info"),
             _("The dynamic SELinux security type tells libvirt to automatically pick a unique label for the guest process and guest image, ensuring total isolation of the guest. (Default)"))

@@ -1649,6 +1650,7 @@ class vmmDetails(vmmGObjectUI):
     def security_type_changed(self, button):
         self.enable_apply(EDIT_SECURITY)
         self.widget("security-label").set_sensitive(not button.get_active())
+        self.widget("security-relabel").set_sensitive(not button.get_active())

     # Memory
     def config_get_maxmem(self):
@@ -1915,13 +1917,15 @@ class vmmDetails(vmmGObjectUI):
             semodel = None
             setype = "static"
             selabel = self.get_text("security-label")
+            relabel = self.widget("security-relabel").get_active()

             if self.widget("security-dynamic").get_active():
                 setype = "dynamic"
+                relabel = True
             if self.widget("security-type-box").get_property("sensitive"):
                 semodel = self.get_text("security-model")

-            add_define(self.vm.define_seclabel, semodel, setype, selabel)
+            add_define(self.vm.define_seclabel, semodel, setype, selabel, relabel)

         if self.editted(EDIT_DESC):
             desc_widget = self.widget("overview-description")
@@ -2472,7 +2476,7 @@ class vmmDetails(vmmGObjectUI):
         self.set_combo_label("overview-clock", clock)

         # Security details
-        semodel, ignore, vmlabel = self.vm.get_seclabel()
+        semodel, sectype, vmlabel, relabel = self.vm.get_seclabel()
         caps = self.vm.conn.get_capabilities()

         if caps.host.secmodel and caps.host.secmodel.model:
@@ -2486,11 +2490,19 @@ class vmmDetails(vmmGObjectUI):
         else:
             self.widget("security-type-box").set_sensitive(bool(semodel))

-            if self.vm.get_seclabel()[1] == "static":
+            if sectype == "static":
                 self.widget("security-static").set_active(True)
+                self.widget("security-relabel").set_sensitive(True)
+                # As "no" is default for relabel with 'static' label and
+                # 'dynamic' must have relabel='yes', this will work properly
+                # for both False (relabel='no') and None (relabel not
+                # specified)
+                self.widget("security-relabel").set_active(relabel)
             else:
                 self.widget("security-dynamic").set_active(True)
-
+                # Dynamic label type must use resource labeling
+                self.widget("security-relabel").set_active(True)
+                self.widget("security-relabel").set_sensitive(False)
             self.widget("security-label").set_text(vmlabel)

     def refresh_stats_page(self):
diff --git a/src/virtManager/domain.py b/src/virtManager/domain.py
index e9d5244..7d3c6c9 100644
--- a/src/virtManager/domain.py
+++ b/src/virtManager/domain.py
@@ -510,13 +510,19 @@ class vmmDomain(vmmLibvirtObject):

     # Security define methods

-    def define_seclabel(self, model, t, label):
+    def define_seclabel(self, model, t, label, relabel):
         def change(guest):
             seclabel = guest.seclabel
             seclabel.model = model or None
             if not model:
                 return

+            if relabel is not None:
+                if relabel:
+                    seclabel.relabel = "yes"
+                else:
+                    seclabel.relabel = "no"
+
             seclabel.type = t
             if label:
                 seclabel.label = label
@@ -901,11 +907,19 @@ class vmmDomain(vmmLibvirtObject):
         return (kernel, initrd, args)

     def get_seclabel(self):
-        model = self._get_guest().seclabel.model
-        t     = self._get_guest().seclabel.type or "dynamic"
-        label = self._get_guest().seclabel.label or ""
+        seclabel = self._get_guest().seclabel
+        model = seclabel.model
+        t     = seclabel.type or "dynamic"
+        label = seclabel.label or ""
+
+        relabel = getattr(seclabel, "relabel", None)
+        if relabel is not None:
+            if relabel == "yes":
+                relabel = True
+            else:
+                relabel = False

-        return [model, t, label]
+        return [model, t, label, relabel]

     # XML Device listing

diff --git a/src/vmm-details.glade b/src/vmm-details.glade
index 0e2d631..7481e06 100644
--- a/src/vmm-details.glade
+++ b/src/vmm-details.glade
@@ -1483,6 +1483,22 @@
                                                         <property name="position">1</property>
                                                       </packing>
                                                     </child>
+                                                    <child>
+                                                      <widget class="GtkCheckButton" id="security-relabel">
+                                                        <property name="label" translatable="yes">relabel</property>
+                                                        <property name="visible">True</property>
+                                                        <property name="can_focus">True</property>
+                                                        <property name="receives_default">False</property>
+                                                        <property name="use_action_appearance">False</property>
+                                                        <property name="draw_indicator">True</property>
+                                                        <signal name="toggled" handler="on_security_relabel_changed" swapped="no"/>
+                                                      </widget>
+                                                      <packing>
+                                                        <property name="expand">False</property>
+                                                        <property name="fill">False</property>
+                                                        <property name="position">2</property>
+                                                      </packing>
+                                                    </child>
                                                   </widget>
                                                   <packing>
                                                     <property name="right_attach">2</property>
@@ -1727,7 +1743,7 @@ I/O:</property>
                                               <widget class="GtkLabel" id="overview-memory-usage-text">
                                                 <property name="visible">True</property>
                                                 <property name="xalign">0</property>
-                                                <property name="label">30 MB of 
+                                                <property name="label">30 MB of
 128 MB</property>
                                               </widget>
                                               <packing>
-- 
1.8.3.2

